clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -440,6 +440,7 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
             correlate(req.request().header(), header);
             if (!metadataUpdater.maybeHandleCompletedReceive(req, now, body))
                 responses.add(new ClientResponse(req, now, false, body));
+            inFlightRequests.completeNext(source);
```

Comment:
@junrao Agreed. Initially I am thinking about letting the Request eventually timeout, we can do that by add a ```failNext()``` method in ```InFlightRequests```. It will remove the request from the in flight requests and time it out (disconnect, etc) immediately. If a response parse failed that essentially means we read some garbage from the socket, hence we'd better close the socket and re-connect. 

We might not be able to return a null body ClientResponse because this has been used by acks=0 case.